hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/EntitySelectFetchInitializerBuilder.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.sql.results.graph.entity.internal;
+
+import org.hibernate.metamodel.mapping.EmbeddableValuedModelPart;
+import org.hibernate.metamodel.mapping.internal.ToOneAttributeMapping;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.spi.NavigablePath;
+import org.hibernate.sql.results.graph.AbstractFetchParentAccess;
+import org.hibernate.sql.results.graph.AssemblerCreationState;
+import org.hibernate.sql.results.graph.DomainResult;
+import org.hibernate.sql.results.graph.FetchParentAccess;
+import org.hibernate.sql.results.graph.embeddable.EmbeddableInitializer;
+
+public class EntitySelectFetchInitializerBuilder {
+
+	public static AbstractFetchParentAccess createInitializer(
+			FetchParentAccess parentAccess,
+			ToOneAttributeMapping fetchedAttribute,
+			EntityPersister entityPersister,
+			DomainResult<?> keyResult,
+			NavigablePath navigablePath,
+			boolean selectByUniqueKey,
+			AssemblerCreationState creationState) {
+		if ( selectByUniqueKey ) {
+			return new EntitySelectFetchByUniqueKeyInitializer(
+					parentAccess,
+					fetchedAttribute,
+					navigablePath,
+					entityPersister,
+					keyResult.createResultAssembler( parentAccess, creationState )
+			);
+		}
+		if ( entityPersister.isBatchLoadable() && !creationState.isScrollResult() ) {
+			if ( parentAccess.isEmbeddableInitializer() ) {
+				if ( isEmbeddedIdPart( parentAccess ) ) {
+					return new EntitySelectFetchInitializer(
+							parentAccess,
+							fetchedAttribute,
+							navigablePath,
+							entityPersister,
+							keyResult.createResultAssembler( parentAccess, creationState )
+					);
+				}
+				return new BatchEntityInsideEmbeddableSelectFetchInitializer(
+						parentAccess,
+						fetchedAttribute,
+						navigablePath,
+						entityPersister,
+						keyResult.createResultAssembler( parentAccess, creationState )
+				);
+			}
+
+			return new BatchEntitySelectFetchInitializer(
+					parentAccess,
+					fetchedAttribute,
+					navigablePath,
+					entityPersister,
+					keyResult.createResultAssembler( parentAccess, creationState )
+			);
+		}
+		else {
+			return new EntitySelectFetchInitializer(
+					parentAccess,
+					fetchedAttribute,
+					navigablePath,
+					entityPersister,
+					keyResult.createResultAssembler( parentAccess, creationState )
+			);
+		}
+	}
+
+	private static boolean isEmbeddedIdPart(FetchParentAccess parentAccess) {
```

Comment:
Let's rename that to e.g. `shouldDisableBatching` and also add checks for the `EmbeddableValuedModelPart` to see if it has a custom instantiator, since when that is present, we can't inject batch loaded values.

Reply:
Not sure if we want to handle this as part of this PR as well, but there are other issues related to batch fetching where we will have to disable batching:

* https://hibernate.atlassian.net/browse/HHH-16025
* https://hibernate.atlassian.net/browse/HHH-16039

Reply:
Also, it would be nice to have a sort of batching implementation where we return a proxy for the initialized instance (if possible) and only batch initialize a these objects in the finishUp method.